### ISSUE TITLE
Express returns a non-compliant HTTP/206 response when gzip is enabled

### ISSUE THREAD

**Issue by mscbpi** [expressjs/compression#185]
CDNs (mostly Azure Front Door) is using [HTTP Range Requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) to retrieve data from Origin when caching enabled.

Express should ignore `Range: ` requests header when `gzip` compression is in place since Express is unable to respond a `HTTP/206` compliant answer with the right computed `Content-Range` header in the response that is taking the compressed data length into account. 

A fair compliant workaround is to, in that very case of compression where computing `Content-Range` values would be too complex, ignore client's `Range:` header in the request, and answer the whole compressed content in a HTTP/200 response.

Handling `Range: ` headers is optional so answering a `HTTP/200` is OK.

Answering a `HTTP/206` with wrong `Content-Range` values is notOK.

Meanwhile another workaround is to disable compression and make CDN handle it, or disable CDN caching, however it would be fair to expect Express to return a compliant HTTP response in any case.

References:

[RFC7233](https://datatracker.ietf.org/doc/html/rfc7233)

Details and highlighting Express behavior:

https://github.com/DanielLarsenNZ/nodejs-express-range-headers

**Comment by IbrahimTanyalcin** [comment]
if the client send a `Range` request header and also `Accept` request header with the `application/gzip` mime type, then it is the clients responsibility to take into account whether the sent byte ranges correspond to zipped file or not. This is the intuitive behavior. 

From the top of my head, I can tell a couple of cases where this change might break applications that depend on genomics data that needs to handle huge files such as https://github.com/igvteam/igv.js/ that make request to [`.bgz`](http://www.htslib.org/doc/bgzip.html), `fasta` or `bam` files. These files arrive to the client compressed, decompressed by the browser and then the client knows that the bytes correspond to the uncompressed payload. 

The decompression is handled by the browser and the final receiver gets the bytes of the payload that corresponds to the uncompressed form. This has always been like this. Am I misinterpreting something?

**Comment by dougwilson** [comment]
Hi @IbrahimTanyalcin you can make any rule you like to determine if the response should br compressed or not using the `filter` option. You can find more and an example at https://github.com/expressjs/compression#filter-1

**Comment by IbrahimTanyalcin** [comment]
@dougwilson I understand that I can pass a filter function, but that means dropping support for compression for plain text file types (fasta etc.) that would actually benefit the greatest from such compression. The benefit from compressing a bundle from Next.js of let's say 300kB is nothing compared to the benefit of compressing a 10Mb chunk of a 5Gb genomic sequence file. Am I wrong in my reasoning?

**Comment by IbrahimTanyalcin** [comment]
Ok here is some more info, in my route file for serving static content, I have something like:
```js
app.use('/static', compression(), _static, function(req, res, next) { //downstream middleware ...
```

`_static` is the expressjs static middleware.

I also dug into what are the request headers for such large fasta files and compared them to regular js/html/css files.  

Here is the causal js/html/css files:
![expressjs-compress-ss1](https://github.com/expressjs/compression/assets/8837895/e1837794-bae5-4a73-b684-da199c71e99f)

And here is the request headers I send for a supposedly large genomic file:
![expressjs-compress-ss3](https://github.com/expressjs/compression/assets/8837895/98585b57-acd1-4326-ac83-b922857f8c75)

So it turns out, the client logic is sending `Accept-Encoding: identity` header which in the [compression library](https://github.com/expressjs/compression/blob/master/index.js#L186-L189)  does NOT compress and is passed to static which doesn't know what to do with `.fasta` file so it adds `Content-Type: application/octet-stream`.

This means in my case this wouldn't seem to break the behavior with large files, as they are already not compressed. I was **wrong**.  ( There might still be other genomics apps though that do not send the correct encoding request header and expect it work though 🤷)

**Comment by IbrahimTanyalcin** [comment]
@mscbpi  yes you are correct, I never objected the RFC, however from a technical standpoint it is very costly to compute the total zipped size and adjust the `Content-Range` value. 

I wish a provisionary header like `Content-Range-Origin: transformed | identity` existed so that the ambiguity is clarified and servers are given a choice. But it is ok, I guess safest bet is to send `Accept-Encoding: identity` or use the `filter` option to drop support for that file type and hope for the best.
